choral/libs/choral-reactive/src/main/java/choral/reactive/ReactiveServer.java

```diff
@@ -58,19 +77,21 @@ public void listen(String address) throws URISyntaxException, IOException {
 
     @SuppressWarnings("unchecked")
     @Override
-    public <T extends Serializable> T recv(Session session) {
+    public <T extends Serializable> T recv(Session session) { // TODO profile this
         Attributes attributes = Attributes.builder().put("channel.service", serviceName)
                 .put("channel.sender", session.senderName()).build();
 
         TelemetrySession telemetrySession;
         CompletableFuture<Serializable> future = new CompletableFuture<Serializable>()
                 .orTimeout(10, TimeUnit.SECONDS);
+                // TODO Do we want this timeout? What happens if the sender takes >10s to compute the message?
```

Comment:
The microservice book says to always use timeouts, but the timeout is essential for failures, since a choreography failing could otherwise result in other roles waiting to receive indefinitely.

Reply:
Aha, I think I understand now. When the future times out, we throw a runtime exception that crashes the virtual thread. This cleans up the future and any resources associated with the session, except for the known_sessions map. Right?

Reply:
Yes :)